Jellyfin.Server.Implementations/MediaSegments/MediaSegmentsManager.cs

```diff
@@ -0,0 +1,213 @@
+                ValidateSegment(segment);
+
+                var found = await dbContext.Segments.FirstOrDefaultAsync(s => s.ItemId.Equals(segment.ItemId) && s.StreamIndex == segment.StreamIndex && s.Type == segment.Type && s.TypeIndex == segment.TypeIndex)
+                    .ConfigureAwait(false);
+
+                AddOrUpdateSegment(dbContext, segment, found);
+            }
+
+            await dbContext.SaveChangesAsync().ConfigureAwait(false);
+        }
+
+        return segments;
+    }
+
+    /// <inheritdoc/>
+    public async Task<IReadOnlyList<MediaSegment>> GetAllMediaSegments(Guid itemId, int? streamIndex = null, int? typeIndex = null, MediaSegmentType? type = null)
+    {
+        var item = _libraryManager.GetItemById(itemId);
+        if (item is null)
+        {
+            throw new InvalidOperationException("Item not found");
+        }
+
+        List<MediaSegment> allSegments;
+
+        var dbContext = await _dbProvider.CreateDbContextAsync().ConfigureAwait(false);
+        await using (dbContext.ConfigureAwait(false))
+        {
+            IQueryable<MediaSegment> queryable = dbContext.Segments.Where(s => s.ItemId.Equals(itemId));
+
+            if (streamIndex is not null)
+            {
+                queryable = queryable.Where(s => s.StreamIndex == streamIndex.Value);
+            }
+
+            if (type is not null)
+            {
+                queryable = queryable.Where(s => s.Type == type.Value);
+            }
+
+            if (!typeIndex.Equals(null))
+            {
+                queryable = queryable.Where(s => s.TypeIndex == typeIndex.Value);
+            }
+
+            allSegments = await queryable.AsNoTracking().ToListAsync().ConfigureAwait(false);
+        }
+
+        return allSegments;
+    }
+
+    /// <summary>
+    /// Add or Update a segment in db.
+    /// <param name="dbContext">The db context.</param>
+    /// <param name="segment">The segment.</param>
+    /// <param name="found">The found segment.</param>
+    /// </summary>
+    private void AddOrUpdateSegment(JellyfinDbContext dbContext, MediaSegment segment, MediaSegment? found)
+    {
+        if (found != null)
+        {
+            found.StartTicks = segment.StartTicks;
+            found.EndTicks = segment.EndTicks;
+            found.Action = segment.Action;
+            found.Comment = segment.Comment;
+        }
+        else
+        {
+            dbContext.Segments.Add(segment);
+        }
+    }
+
+    /// <summary>
+    /// Validate a segment: itemId, start >= end and type.
+    /// </summary>
+    /// <param name="segment">The segment to validate.</param>
+    private void ValidateSegment(MediaSegment segment)
+    {
+        if (segment.ItemId.Equals(default))
+        {
+            throw new ArgumentException($"itemId is default: itemId={segment.ItemId} for segment with type '{segment.Type}.{segment.TypeIndex}'");
+        }
+
+        if (segment.StartTicks >= segment.EndTicks)
```

Comment:
You can use the `ArgumentOutOfRangeException.ThrowIfGreaterThanOrEqual` helper here.